File: Python/Partie7/Exo29.py
import os, signal, time, sys

fils = []
for _ in range(5):
    pid = os.fork()
    if pid == 0:
        def term(sig, frame):
            print(f"Fils {os.getpid()} terminé")
            sys.exit(0)
        signal.signal(signal.SIGTERM, term)
        while True:
            time.sleep(1)
    else:
        fils.append(pid)

def pere_term(sig, frame):
    print("Père tué")
    sys.exit(0)
signal.signal(signal.SIGTERM, pere_term)

while True:
    print("1-Endormir, 2-Réveiller, 3-Terminer, 4-Quitter")
    choix = int(input())
    if choix == 4:
        break
    num = int(input("Numéro (1-5) : ")) - 1
    if 0 <= num < 5:
        if choix == 1:
            os.kill(fils[num], signal.SIGSTOP)
        elif choix == 2:
            os.kill(fils[num], signal.SIGCONT)
        elif choix == 3:
            os.kill(fils[num], signal.SIGTERM)
for p in fils:
    os.kill(p, signal.SIGTERM)
for _ in range(5):
    os.wait()

File: C/Partie7/Exo29.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

void fils_mourant(int sig)
{
    printf("Fils %d terminé\n", getpid());
    exit(0);
}

int main()
{
    pid_t fils[5];
    for (int i = 0; i < 5; i++)
    {
        fils[i] = fork();
        if (fils[i] == 0)
        {
            signal(SIGTERM, fils_mourant);
            while (1)
                pause();
        }
    }
    int choix, num;
    while (1)
    {
        printf("1-Endormir, 2-Réveiller, 3-Terminer, 4-Quitter\n");
        scanf("%d", &choix);
        if (choix == 4)
            break;
        printf("Numéro du fils (1-5) : ");
        scanf("%d", &num);
        if (num < 1 || num > 5)
            continue;
        if (choix == 1)
            kill(fils[num - 1], SIGSTOP);
        else if (choix == 2)
            kill(fils[num - 1], SIGCONT);
        else if (choix == 3)
            kill(fils[num - 1], SIGTERM);
    }
    for (int i = 0; i < 5; i++)
        kill(fils[i], SIGTERM);
    while (wait(NULL) > 0)
        ;
    return 0;
}